stream_alert/shared/publisher.py

```diff
@@ -0,0 +1,272 @@
+            try:
+                publication = deepcopy(publication)
+                publication = publisher.publish(alert, publication)
+            except KeyError:
+                LOGGER.exception(
+                    'CompositePublisher encountered KeyError with publisher: %s',
+                    publisher.__name__
+                )
+                raise
+
+        return publication
+
+
+class WrappedFunctionPublisher(AlertPublisher):
+    """A class only used to wrap a function publisher."""
+
+    def __init__(self, function):
+        self._function = function
+
+    def publish(self, alert, publication):
+        return self._function(alert, publication)
+
+
+class AlertPublisherRepository(object):
+    """A repository mapping names -> publishers
+
+    As a usability optimization, using this Repository will eagerly load and register all
+    publishers in the application.
+    """
+    _publishers = {}
+
+    @staticmethod
+    def is_valid_publisher(thing):
+        """Returns TRUE if the given reference can be registered as a publisher
+
+        Publishers are valid if and only if they fall into one of the following categories:
+
+        * They are a python function that accepts 2 arguments: (alert: Alert, publication: dict)
+        * They are a python class that extends AlertPublisher
+
+        Args:
+            thing (mixed): Any primitive or reference to be checked
+
+        Returns:
+            bool
+        """
+
+        # We have to put the isclass() check BEFORE the callable() check because classes are also
+        # callable!
+        if isclass(thing) and issubclass(thing, AlertPublisher):
+            return True
+        elif callable(thing):
+            return True
+
+        return False
+
+    @staticmethod
+    def get_publisher_name(class_or_function):
+        """Given a class or function, will return its fully qualified name.
+
+        This is useful for assigning a unique string name for a publisher.
+
+        Args:
+            class_or_function (callable|Class): A reference to a python function or class
+
+        Returns:
+            string
+        """
+        return '{}.{}'.format(class_or_function.__module__, class_or_function.__name__)
+
+    @classmethod
+    def register_publisher(cls, publisher):
+        """Registers the publisher into the repository.
+
+        To standardize the interface of publishers, if a function publisher is given, it will be
+        wrapped with a WrappedFunctionPublisher instance prior to being registed into the
+        Repository.
+
+        Args:
+             publisher (callable|AlertPublisher): An instance of a publisher class or a function
+        """
+        if not AlertPublisherRepository.is_valid_publisher(publisher):
+            LOGGER.error(
+                'Could not register publisher %s; Not callable nor subclass of AlertPublisher',
+                publisher
+            )
+            raise PublisherRegistrationError
```

Comment:
custom exceptions that are raised should be instantiated (preferably with a helpful message). In this case, since you are logging an error and then raising an exception, I'd just suggest not logging the error and using that error message when raising the exception. but.. do the message formatting outside of the exception otherwise it looks ugly. so:

```python
error = 'Could not register publisher {}; Not callable nor subclass of AlertPublisher'.format(publisher)
raise PublisherRegistrationError(error)
```

Reply:
example of way to _not_ do the formatting:
```
Query '<id>' FAILED with reason 'FAILED: SemanticException Unable to fetch table cloudwatch_cloudtrail. null (Service: AmazonDataCatalog; Status Code: 500; Error Code: ServiceUnavailableException; Request ID: <id>)', exiting: AthenaQueryExecutionError
Traceback (most recent call last):
File "/var/task/stream_alert/athena_partition_refresh/main.py", line 219, in handler
AthenaRefresher().run(event)
File "/var/task/stream_alert/athena_partition_refresh/main.py", line 211, in run
if not self._add_partitions():
File "/var/task/stream_alert/athena_partition_refresh/main.py", line 166, in _add_partitions
success = self._ATHENA_CLIENT.run_query(query=query)
File "/var/task/stream_alert/shared/athena.py", line 270, in run_query
return bool(self._execute_and_wait(query))
File "/var/task/stream_alert/shared/athena.py", line 106, in _execute_and_wait
self.check_query_status(exeuction_id)
File "/var/task/stream_alert/shared/athena.py", line 218, in check_query_status
'Query \'
{}
\'
{}
with reason \'
{}
\', exiting'.format(execution_id, state, reason)
AthenaQueryExecutionError: Query '<id>' FAILED with reason 'FAILED: SemanticException Unable to fetch table cloudwatch_cloudtrail. null (Service: AmazonDataCatalog; Status Code: 500; Error Code: ServiceUnavailableException; Request ID: <id>)', exiting
```